ui/org.eclipse.pde.junit.runtime/src/org/eclipse/pde/internal/junit/runtime/RemotePluginTestRunner.java

```diff
@@ -99,19 +104,38 @@ private static ClassLoader createJUnit5PluginClassLoader(String testPluginName)
 
 	private static List<Bundle> findTestEngineBundles() {
 		BundleContext bundleContext = FrameworkUtil.getBundle(RemotePluginTestRunner.class).getBundleContext();
-		List<Bundle> engineBundles = new ArrayList<>();
-		for (Bundle bundle : bundleContext.getBundles()) {
-			try {
-				BundleWiring bundleWiring = bundle.adapt(BundleWiring.class);
-				Collection<String> listResources = bundleWiring.listResources("META-INF/services", "org.junit.platform.engine.TestEngine", BundleWiring.LISTRESOURCES_LOCAL); //$NON-NLS-1$//$NON-NLS-2$
-				if (!listResources.isEmpty()) {
-					engineBundles.add(bundle);
-				}
-			} catch (Exception e) {
-				// check the next bundle
+		return Arrays.stream(bundleContext.getBundles()).filter(RemotePluginTestRunner::providesCompatibleTestEngine).collect(toList());
+	}
+
+	/**
+	 * Checks whether the bundle provides test engines.
+	 * Ensures that test engines that can be loaded from the bundle
+	 * are compatible with the TestEngine version in current scope.
+	 * Otherwise, the JUnit platform's call to the ServiceLoader for
+	 * retrieving available engines will fail.
+	 * Incompatibilities can happen, e.g., in Tycho builds, where
+	 * the org.eclipse.tycho.surefire.osgibooter bundle is found
+	 * that may provide a different JUnit platform version than the
+	 * one available via the Eclipse target platform.
+	 */
+	private static boolean providesCompatibleTestEngine(Bundle bundle) {
+		try {
+			BundleWiring bundleWiring = bundle.adapt(BundleWiring.class);
+			Collection<String> engineProviderSpecifications = bundleWiring.listResources("META-INF/services", "org.junit.platform.engine.TestEngine", BundleWiring.LISTRESOURCES_LOCAL); //$NON-NLS-1$//$NON-NLS-2$
+			if (!engineProviderSpecifications.isEmpty()) {
+				Bundle engineBundleInThisClassLoader = FrameworkUtil.getBundle(TestEngine.class);
+				Class<?> testEngineClassInCurrentLoader = getTestEngineClassInBundle(engineBundleInThisClassLoader);
+				Class<?> testEngineInEngineProviderBundle = getTestEngineClassInBundle(bundle);
+				return testEngineInEngineProviderBundle != null && testEngineClassInCurrentLoader.isAssignableFrom(testEngineInEngineProviderBundle);
```

Comment:
In order to avoid a hard dependency on `org.junit.platform.engine` I suggest using the following code.

```suggestion
			String testEngineClass = "org.junit.platform.engine.TestEngine"; //$NON-NLS-1$
			Collection<String> engineProviders = bundleWiring.listResources("META-INF/services", testEngineClass, BundleWiring.LISTRESOURCES_LOCAL); //$NON-NLS-1$
			if (!engineProviders.isEmpty()) {
				Class<?> thisTestEngine = Class.forName(testEngineClass);
				Class<?> bundleTestEngine = bundle.loadClass(testEngineClass);
				return bundleTestEngine != null && thisTestEngine.isAssignableFrom(bundleTestEngine);
			}
```

I also suggest to shorten the variable names a bit. While descriptive names are definitively important, too long names are IMHO badly readable as well.

Reply:
Incorporated in f4e9f3358757e62cc21c0e9e0be12a0baac96a31